### ISSUE TITLE
STUN re-discovery not happening after suspend/sleep of host

### ISSUE THREAD

**Issue by omgold** [syncthing/syncthing#8341]
I run Syncthing 1.20.1 on Linux.

I notice, when I use quic with NAT traversal via STUN, that after putting the host into suspend mode (S3) for a prolonged time, no new quic connections are established. When I look for the entry on the discosrv, the host inserts again the previous remote/NAT'ed IP/port which it had before the suspend. More often than not these are not correct anymore, of course.

Looking around a bit in the code, I understand the logic here, such that no STUN servers are contacted again, as long as there is traffic on the local port. As the peer still has a quic address advertised, there will be (only outgoing) traffic to try to reestablish a connection. Thus the discovery address is never updated and the firewall of the peer remains closed, as it tries to contact only the old address.

**Comment by bt90** [comment]
This sounds like a duplicate of #8243

**Comment by AudriusButkevicius** [comment]
Thanks, its now clear what the issue is, but we usually restart after detecting sleep to mitigate stuff like this, is that not working?

**Comment by omgold** [comment]
Doesn't look like. What exactly is supposed to be restarted automatically? The whole process?

**Comment by AudriusButkevicius** [comment]
Yes, thats the default.

**Comment by omgold** [comment]
Okay, seems I found the issue, then:

The code to detect the restart is this

```
			if time.Since(now) > 2*time.Minute && cfg.Options().RestartOnWakeup {
			l.Infof("Paused state detected, possibly woke up from standby. Restarting in %v.", restartDelay)
```

But the manual for the go time packages says this:


_Operating systems provide both a “wall clock,” which is subject to changes for clock synchronization, and a “monotonic clock,” which is not. The general rule is that the wall clock is for telling time and the monotonic clock is for measuring time. Rather than split the API, in this package the Time returned by time.Now contains both a wall clock reading and a monotonic clock reading; later time-telling operations use the wall clock reading, but later time-measuring operations, specifically comparisons and subtractions, use the monotonic clock reading._

And indeed, when I print out the values like this:

```
		l.Infof("standby monitor: before: %s, after: %s, diff: %s, wakeup: %s, opt: %s", now, time.Now(), time.Since(now), time.Since(now) > 2*time.Minute, cfg.Options().RestartOnWakeup)
```

I get:

```
[H4352] INFO: standby monitor: before: 2022-05-14 07:55:30.73093779 +0200 CEST m=+140.036064024, after: 2022-05-14 07:58:54.876967931 +0200 CEST m=+152.102880692, diff: 12.066816719s, wakeup: %!s(bool=false), opt: %!s(bool=true)
```

In short, `time.Since()` is not the right thing to measure differences in walltime.

But back to the STUN issue. In my experience with NAT traversal, it is far from guaranteed that mapping do not change, even with the per-15s keepalive pings the quic implementation does. So IMHO, occasionally doing STUN re-discovery even when a connection is alive would be a good thing to have.

**Comment by AudriusButkevicius** [comment]
Wtf? How is that 12 seconds when the wallclock time is 2 mins?

**Comment by omgold** [comment]
@AudriusButkevicius The quote from the manual explains it. `time.Since()` does not calculate the difference in wallclock time, but monotonic time. And monotonic time is apparently halted during system suspend.

**Comment by AudriusButkevicius** [comment]
I get that, I just find it very surprising.

**Comment by omgold** [comment]
Yeah, quite a bit counterintuitive.

The fix is simple, though:

```
--- src/github.com/syncthing/syncthing/cmd/syncthing/main.go.orig	2022-05-14 15:13:31.192225539 +0200
+++ src/github.com/syncthing/syncthing/cmd/syncthing/main.go	2022-05-14 15:13:52.694070436 +0200
@@ -763,7 +763,7 @@
 
 func standbyMonitor(app *syncthing.App, cfg config.Wrapper) {
 	restartDelay := 60 * time.Second
-	now := time.Now()
+	now := time.Now().Round(0)
 	for {
 		time.Sleep(10 * time.Second)
 		if time.Since(now) > 2*time.Minute && cfg.Options().RestartOnWakeup {
@@ -777,7 +777,7 @@
 			app.Stop(svcutil.ExitRestart)
 			return
 		}
-		now = time.Now()
+		now = time.Now().Round(0)
 	}
 }
 ```

**Comment by AudriusButkevicius** [comment]
Feel free to open a PR given you already got a diff.
